### DISCUSSION TITLE
Budget summary boxes miss periods sometimes

### DISCUSSION THREAD

**Discussion by likuilin** [firefly-iii/firefly-iii#7733] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

If I create a new budget, and then set a lot of pre-existing transactions (before the current budget period) to that budget, I don't see those periods in the aggregate boxes on the right of the budget page.

### Debug information

The /debug for the demo page doesn't work, and I think would be unhelpful anyways.

### Expected behaviour

My use case is that I would like to retroactively re-categorize which expenses are part of which budgets, and then see the effect of that re-allocation. In doing that, I made a new budget, but I was expecting to be able to see the past monthly amounts that the budget would have had.

### Steps to reproduce

I was able to reproduce this on the demo site just now by doing:

* Create a new budget, setting the budget name and an allocation for the current month
* Set all transactions with destination "Greenchoice Electricity" to new budget with bulk edit
* Set a bunch of other transactions in the current month to the new budget with bulk edit
* Set all transactions in April (before the current month) to the new budget with bulk edit

Now, the new budget's page has a summary box for July, but not one for April, which I'm expecting. I'm not sure how often the demo site resets, but here: https://demo.firefly-iii.org/budgets/show/6/85

**Comment by likuilin** [comment]
I managed to fix this, I think, by adding an allocation of $1 to a missing period for the budget, and then bulk re-editing the transactions without changing anything. However, now, changing the budget allocation back to $0, or $2 for the matter, for that period, doesn't even update the boxes.

Is it intended that if a budget is allocated $0, the box shouldn't show up? I think it should, because the Spent field is still useful.

Edit: I got the data I needed by duplicating my instance, assigning $1 to every period from when I started using Firefly to today, for the budget, and then bulk re-editing all my transactions. I don't want to keep the $1 assigned to each period, because it should be $0 to preserve the accuracy of that historical data, hence the duplicate instance.

**Comment by JC5** [comment]
You have discovered a feature, not a bug. The boxes in the side bar of a budget are drawn according to the amount of money assigned to periods. 

If you want historical data for a budget, there has to be money budgeted. Use a category if this isn't what you need.

**Reply by likuilin** [reply]
The boxes not being re-calculated if the amount of money assigned to them changes is a bug, though, right?

**Reply by JC5** [reply]
What happens after you browse to /flush ?

**Reply by likuilin** [reply]
Doesn't flush it. I just tested on the demo site:

* Browse to Groceries budget for June 2023
* Change allocation from €100 to €5
* Visit budget page, see €100 in right side box for June 2023
* Visit /flush, hard refresh, etc
* Visit budget page, see €100 in right side box for June 2023
* Edit a random transaction during June on the Groceries budget and save without changing anything
* Visit budget page, see €5 in right side box for June 2023

**Reply by JC5** [reply]
That still implies it's the cache 🤔. I'll check it out.

**Reply by JC5** [reply]
It's a cache thing. Not sure why /flush doesn't work, I'll check it out 👍🏻 

Edit: should be fixed in the next release
